src/sensitivities/chainrules.jl

```diff
@@ -0,0 +1,197 @@
+#using InteractiveUtils
+
+function generate_overload(sig)
+    opT, argTs = Iterators.peel(ExprTools.parameters(sig))
+    opT <: Core.Builtin && return false  # can't do operater overloading for builtins
+
+    isabstracttype(opT) || fieldcount(opT) == 0 || return false # not handling functors
+    isempty(argTs) && return false  # we are an operator overloading AD, need operands
+
+    opT isa DataType && nameof(opT.name.module) == :NaNMath  && return false # Don't care about NaNMath
+
+    # Ignore functions that have complex ranges. This may change when Nabla supports complex
+    # numbers.
+    opT ∈ typeof.((
+        SpecialFunctions.hankelh1, SpecialFunctions.hankelh2,
+        log1p, rem2pi, mod, atan, rem,
+    ))  && return false
+    opT <: Type{<:Complex} && return false  # skip complex constructor
+
+    # Ignore these functions because they have better Nabla specific versions.
+    opT ∈ typeof.((
+        isapprox, size, length, isassigned,
+        Base.Broadcast.combine_styles,  #TODO should i keep this?
+    )) && return false
+
+
+    signature_def = build_def(sig)
+    original_signature_args = signature_def[:args]
+    signature_def[:args] = unionise_sig.(original_signature_args)
+
+    fdef = quote
```

Comment:
With all of these different cases, I could see an increase in readability if they belonged to their own functions.